packages/tasit-contracts/decentraland/scripts/3_mana_token.js

```diff
@@ -0,0 +1,5 @@
+const MANAToken = artifacts.require("./MANAToken.sol");
```

Comment:
For SQL migrations the rule is that each integer prefix for migration file names should only be used once. Do you think that's the case for Truffle too? I would guess so.

Reply:
In that case, I think these 3 migrations could move into one file since they all came in one PR

Reply:
A) When `truffle migrate` runs that will execute all `js` files inside `migration` following the prefix number order. The `Migrations.last_completed_migration` will store the number/prefix of the last script executed and prevent that already deployed contract be deployed again. When a new contract and a new migration script are created, `truffle migration` will deploy only that new contract.
Note: `truffle migrate --reset` will ignore already deployed contracts and re-redeploy all.

Reply:
B) All of these projects (mana, land and marketplace-contracts) use a shell script for deploy'n'tests, and all have the default `1_initial_migration.js` file and an empty `2_deploy_contracts.js`. The idea was not to change 3rd parties repos, because of that I've created these `3` scripts, one for each repo, to be copied to each repo and allows `truffle migrate` to work.

Reply:
A) Makes sense, yes, this is my understanding.

B) Ah ok, that makes sense. Just to consider all possibilities, could we have used their internal deployment scripts without adding our own migration at all? Or would that have been harder? I guess if they weren't using truffle or `ganache-cli` but we wanted to still for that project that could get very complicated quickly, and would warrant our own deployment strategy.

Reply:
I'll paste the script here and create a new comment:

```sh
#!/usr/bin/env bash

# Exit script as soon as a command fails.
set -o errexit

# Executes cleanup function at script exit.
trap cleanup EXIT

cleanup() {
  # Kill the ganache instance that we started (if we started one and if it's still running).
  if [ -n "$ganache_pid" ] && ps -p $ganache_pid > /dev/null; then
    kill -9 $ganache_pid
  fi
}

if [ "$SOLIDITY_COVERAGE" = true ]; then
  ganache_port=8555
else
  ganache_port=8545
fi

ganache_running() {
  nc -z localhost "$ganache_port"
}

start_ganache() {
  # We define 10 accounts with balance 1M ether, needed for high-value tests.
  local accounts=(
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501200,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501201,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501202,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501203,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501204,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501205,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501206,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501207,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501208,1000000000000000000000000"
    --account="0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501209,1000000000000000000000000"
  )

  node_modules/.bin/ganache-cli --gasLimit 0xfffffffffff "${accounts[@]}" --port "$ganache_port" > /dev/null &

  ganache_pid=$!
}

if ganache_running; then
  echo "Using existing ganache instance"
else
  echo "Starting our own ganache instance"
  start_ganache
fi

node_modules/.bin/truffle test "$@"
```

Reply:
> B) Ah ok, that makes sense. Just to consider all possibilities, could we have used their internal deployment scripts without adding our own migration at all? Or would that have been harder? I guess if they weren't using truffle or ganache-cli but we wanted to still for that project that could get very complicated quickly, and would warrant our own deployment strategy.

I've just run the script here with our ganache running and that works. My only note is that has only test script that makes migrate too, will not have a `just migrate` script for these projects. 
But it's doesn't seems a problem, I could:
1) Remove tests from `tasit-contract`
2) Since `tasit-contract` tests runs before `tasit-action` tests these contract will be deployed just in time.
3) Get abi from decentraland projects folders.

Sounds good?

Reply:
That sounds good to me if that sounds good to you! I'm still open to the earlier approach if you prefer it.

Reply:
I think that minimizing touching 3rd parties projects is better. I'll go on that way and see if it'll work.

Reply:
When `truffle test` is called without migrate (same as the script does above), it deploys the required contracts but only for tests and uses snapshots to clear the blockchain when test ends. Because of that, this script in fact doesn't do the same as `truffle migrate` does. Any idea?

Reply:
Maybe an intermediary solution:
Still pasting migration js files into projects but using their scripts to test them instead of having their tests inside `tasit-contract`.

Reply:
I've thinked a little more and I want to share some points:
- We really need run 3th part intire test suite? It will increate our `lerna test` runtime;
- We could be exposed to 3th part non-deterministic tests (I think that isn't the case of Decentraland);
- I like the idea to have at least one test case for each 3th part contract, for now it's useful to check if the address for each one is correct.

Reply:
Hmm yeah, so what we really want to run is the local migration (as opposed to testing) script for the 3rd-party contracts. I'll have to look whether enough teams expose a migration script that we can count on that being present (and there's still a risk that they'll use the same defaults and compete with each other for a single ganache instance or something).

I agree that running their whole test suite is probably overkill / risky from a CI perspective except once each time we update their contract versions.

Yeah, one test case for each makes sense. The more the test case is reusable or DRY between different projects, the less work it will be to add it each time.

Reply:
So given that there's still some more thinking to do here, I think this is okay to merge in its current state. We can track the "how best to run migrations for 3rd-party contracts" in a separate issue.

Reply:
Thanks for the quick response! So, I'll go on with my first approach, alright?

Reply:
OK!

Reply:
https://github.com/tasitlabs/TasitSDK/issues/154

Reply:
Thanks. Resolving!